File: src/parser/pyParser.py
import ast
import json
import sys
import os
import copy

class PyBodyStripper(ast.NodeTransformer):
    def visit_FunctionDef(self, node):
        node.body = [ast.Pass()]
        return node
    def visit_AsyncFunctionDef(self, node):
        node.body = [ast.Pass()]
        return node

def parse_python_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        code = f.read()

    try:
        tree = ast.parse(code, filename=file_path)
    except Exception as e:
        return {"error": str(e), "filePath": file_path, "imports": [], "symbols": []}

    imports = []
    symbols = []

    class DependencyExtractor(ast.NodeVisitor):
        def __init__(self, exclude_name):
            self.exclude_name = exclude_name
            self.names = set()

        def visit_Name(self, node):
            if isinstance(node.ctx, ast.Load) and node.id != self.exclude_name:
                self.names.add(node.id)
            self.generic_visit(node)

        def visit_Attribute(self, node):
            if isinstance(node.value, ast.Name):
                self.names.add(f"{node.value.id}.{node.attr}")
            else:
                self.generic_visit(node)

    for node in tree.body:
        # 1. Extract Imports
        if isinstance(node, ast.Import):
            for name in node.names:
                imports.append({
                    "source": name.name,
                    "specifiers": [{"localName": name.asname or name.name.split('.')[0], "exportName": "*"}]
                })
        elif isinstance(node, ast.ImportFrom):
            if node.module or node.level > 0:
                specifiers = [{"localName": n.asname or n.name, "exportName": n.name} for n in node.names]
                mod_name = node.module or ""
                imports.append({
                    "source": "." * (node.level or 0) + mod_name,
                    "specifiers": specifiers
                })

        # 2. Extract Top-level Symbols (functions and classes)
        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)):
            name = node.name
            symbol_type = "function" if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) else "class"
            
            start_line = node.lineno
            end_line = getattr(node, 'end_lineno', start_line + 10)
            
            lines = code.split('\n')[start_line-1:end_line]
            symbol_code = '\n'.join(lines)

            # Extract dependencies first
            extractor = DependencyExtractor(name)
            extractor.visit(node)

            # Strip body for declaration code
            decl_code = symbol_code
            try:
                # Deepcopy node so we do not mutate the main tree if it's used elsewhere
                node_copy = copy.deepcopy(node)
                stripped_node = PyBodyStripper().visit(node_copy)
                decl_code = ast.unparse(stripped_node)
            except Exception as e:
                # Fallback to symbol_code
                pass

            symbols.append({
                "name": name,
                "type": symbol_type,
                "start": start_line,
                "end": end_line,
                "code": symbol_code,
                "declCode": decl_code,
                "dependencies": list(extractor.names)
            })

        # 3. Extract Top-level Assignments (constants, globals, annotations)
        elif isinstance(node, ast.Assign):
            names_to_add = []
            for target in node.targets:
                if isinstance(target, ast.Name):
                    names_to_add.append(target.id)
                elif isinstance(target, (ast.Tuple, ast.List)):
                    for elt in target.elts:
                        if isinstance(elt, ast.Name):
                            names_to_add.append(elt.id)
            
            for name in names_to_add:
                start_line = node.lineno
                end_line = getattr(node, 'end_lineno', start_line)
                
                lines = code.split('\n')[start_line-1:end_line]
                symbol_code = '\n'.join(lines)
                
                extractor = DependencyExtractor(name)
                extractor.visit(node.value)
                
                symbols.append({
                    "name": name,
                    "type": "other",
                    "start": start_line,
                    "end": end_line,
                    "code": symbol_code,
                    "declCode": symbol_code,
                    "dependencies": list(extractor.names)
                })

        elif isinstance(node, ast.AnnAssign):
            if isinstance(node.target, ast.Name):
                name = node.target.id
                start_line = node.lineno
                end_line = getattr(node, 'end_lineno', start_line)
                
                lines = code.split('\n')[start_line-1:end_line]
                symbol_code = '\n'.join(lines)
                
                extractor = DependencyExtractor(name)
                if node.value:
                    extractor.visit(node.value)
                
                symbols.append({
                    "name": name,
                    "type": "other",
                    "start": start_line,
                    "end": end_line,
                    "code": symbol_code,
                    "declCode": symbol_code,
                    "dependencies": list(extractor.names)
                })

    return {
        "filePath": file_path,
        "imports": imports,
        "symbols": symbols
    }

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No file path provided"}))
        sys.exit(1)
    
    file_path = sys.argv[1]
    if not os.path.exists(file_path):
        print(json.dumps({"error": "File not found"}))
        sys.exit(1)

    result = parse_python_file(file_path)
    print(json.dumps(result))

File: src/parser/pyParser.ts
import * as path from 'path';
import * as fs from 'fs';
import { FileDependencies, SymbolInfo, ImportInfo } from './tsParser';

/**
 * Resolves local python import paths.
 */
export function resolvePyImportPath(importingFilePath: string, source: string): string | null {
  const dir = path.dirname(importingFilePath);
  
  // Handle relative imports e.g. .utils or ..helpers
  if (source.startsWith('.')) {
    let cleanSource = source;
    let currentDir = dir;
    
    // Resolve relative dots
    while (cleanSource.startsWith('.')) {
      cleanSource = cleanSource.slice(1);
      if (cleanSource.startsWith('.')) {
        currentDir = path.dirname(currentDir);
      }
    }
    
    if (cleanSource.length === 0) {
      return currentDir;
    }
    
    const relativeTarget = cleanSource.replace(/\./g, '/');
    const potentialPaths = [
      path.resolve(currentDir, relativeTarget + '.py'),
      path.resolve(currentDir, path.join(relativeTarget, '__init__.py')),
      path.resolve(currentDir, relativeTarget)
    ];

    for (const p of potentialPaths) {
      if (fs.existsSync(p)) {
        return p;
      }
    }
  }

  // Handle absolute imports that might be local files in the project
  const localTarget = source.replace(/\./g, '/');
  const potentialPaths = [
    path.resolve(dir, localTarget + '.py'),
    path.resolve(dir, path.join(localTarget, '__init__.py')),
    path.resolve(dir, localTarget),
    path.resolve(process.cwd(), localTarget + '.py'),
    path.resolve(process.cwd(), path.join(localTarget, '__init__.py')),
    path.resolve(process.cwd(), localTarget)
  ];

  for (const p of potentialPaths) {
    if (fs.existsSync(p)) {
      return p;
    }
  }

  return null;
}

/**
 * Checks if a line is empty or just a comment.
 */
function isLineEmptyOrComment(line: string): boolean {
  const trimmed = line.trim();
  return trimmed.length === 0 || trimmed.startsWith('#');
}

/**
 * Gets the indentation level (number of leading spaces/tabs) of a line.
 */
function getIndentationLevel(line: string): number {
  const match = line.match(/^([ \t]*)/);
  if (!match) return 0;
  let score = 0;
  for (const char of match[1]) {
    if (char === '\t') score += 4;
    else score += 1;
  }
  return score;
}

/**
 * Parses a Python file in pure TypeScript.
 * Uses regular expressions and indentation-based scanning to extract imports and top-level symbols.
 */
export function parsePythonFile(filePath: string): FileDependencies {
  const absolutePath = path.resolve(filePath);
  const code = fs.readFileSync(absolutePath, 'utf-8');
  const lines = code.split('\n');

  const imports: ImportInfo[] = [];
  const symbols: SymbolInfo[] = [];

  // Track local names of imported packages/modules to distinguish them
  const importLocalNames = new Set<string>();

  // 1. First Pass: Parse Imports and collect local import names
  const importRegex = /^\s*import\s+([a-zA-Z0-9_.,\s]+)/;
  const fromImportRegex = /^\s*from\s+([a-zA-Z0-9_.]+)\s+import\s+([a-zA-Z0-9_.,\s()]+)/;

  for (const line of lines) {
    // Match "import X" or "import X as Y"
    const importMatch = line.match(importRegex);
    if (importMatch) {
      const parts = importMatch[1].split(',');
      for (const part of parts) {
        const subParts = part.trim().split(/\s+as\s+/);
        const source = subParts[0].trim();
        const alias = subParts[1] ? subParts[1].trim() : source;
        importLocalNames.add(alias);

        const resolvedPath = resolvePyImportPath(absolutePath, source);
        if (resolvedPath && fs.statSync(resolvedPath).isFile()) {
          imports.push({
            source,
            resolvedPath,
            specifiers: [{ localName: alias, exportName: '*' }]
          });
        }
      }
      continue;
    }

    // Match "from X import Y, Z"
    const fromImportMatch = line.match(fromImportRegex);
    if (fromImportMatch) {
      const source = fromImportMatch[1].trim();
      const importList = fromImportMatch[2].replace(/[()]/g, '');
      const parts = importList.split(',');
      
      const parsedSpecifiers = parts.map(part => {
        const subParts = part.trim().split(/\s+as\s+/);
        const name = subParts[0].trim();
        const alias = subParts[1] ? subParts[1].trim() : name;
        return { localName: alias, exportName: name };
      }).filter(s => s.localName.length > 0);

      // Resolve base path
      const baseResolved = source === '.' ? path.dirname(absolutePath) : resolvePyImportPath(absolutePath, source);

      if (baseResolved) {
        const parentDir = fs.statSync(baseResolved).isDirectory() ? baseResolved : path.dirname(baseResolved);
        
        const submoduleSpecifiers: typeof parsedSpecifiers = [];
        const normalSpecifiers: typeof parsedSpecifiers = [];

        for (const spec of parsedSpecifiers) {
          importLocalNames.add(spec.localName);
          
          const potentialFile = path.resolve(parentDir, spec.exportName + '.py');
          const potentialInitFile = path.resolve(parentDir, spec.exportName, '__init__.py');
          
          let resolvedSub: string | null = null;
          if (fs.existsSync(potentialFile) && fs.statSync(potentialFile).isFile()) {
            resolvedSub = potentialFile;
          } else if (fs.existsSync(potentialInitFile) && fs.statSync(potentialInitFile).isFile()) {
            resolvedSub = potentialInitFile;
          }

          if (resolvedSub) {
            const subSource = source === '.' ? '.' + spec.exportName : source + '.' + spec.exportName;
            imports.push({
              source: subSource,
              resolvedPath: resolvedSub,
              specifiers: [{ localName: spec.localName, exportName: '*' }]
            });
          } else {
            normalSpecifiers.push(spec);
          }
        }

        if (normalSpecifiers.length > 0 && fs.statSync(baseResolved).isFile()) {
          imports.push({
            source,
            resolvedPath: baseResolved,
            specifiers: normalSpecifiers
          });
        }
      }
    }
  }

  // Helper to extract identifiers for dependencies, respecting namespace property access
  function getIdentifiers(symbolCode: string): string[] {
    const refs = new Set<string>();
    
    // Match namespace attribute access e.g., u.hash_password or my_math.add_numbers
    const attrRegex = /[a-zA-Z_][a-zA-Z0-9_]*(?:\.[a-zA-Z_][a-zA-Z0-9_]*)*/g;
    const found = symbolCode.match(attrRegex) || [];
    
    for (const ref of found) {
      const parts = ref.split('.');
      if (parts.length > 1) {
        let isImportNs = false;
        for (let i = 1; i <= parts.length - 1; i++) {
          const prefix = parts.slice(0, i).join('.');
          if (importLocalNames.has(prefix)) {
            isImportNs = true;
            refs.add(parts.slice(0, i + 1).join('.')); // Add namespace + property, e.g. "u.hash_password"
            break;
          }
        }
        if (isImportNs) continue;
      }
      
      refs.add(parts[0]);
    }
    
    return Array.from(refs);
  }

  // 2. Second Pass: Extract top-level symbols
  let charOffset = 0;

  for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
    const line = lines[lineIndex];
    const lineLength = line.length + 1; // +1 for newline

    // Skip import lines during symbol extraction
    if (line.match(importRegex) || line.match(fromImportRegex)) {
      charOffset += lineLength;
      continue;
    }

    const baseIndent = getIndentationLevel(line);
    
    if (baseIndent === 0 && !isLineEmptyOrComment(line)) {
      const defMatch = line.match(/^(async\s+)?def\s+([a-zA-Z0-9_]+)\s*\(/);
      const classMatch = line.match(/^class\s+([a-zA-Z0-9_]+)/);
      const assignMatch = line.match(/^([a-zA-Z0-9_]+)\s*(?::\s*[a-zA-Z0-9_.[\]]+)?\s*=\s*/);

      if (defMatch || classMatch || assignMatch) {
        const name = defMatch ? defMatch[2] : (classMatch ? classMatch[1] : assignMatch![1]);
        const type: SymbolInfo['type'] = defMatch ? 'function' : (classMatch ? 'class' : 'variable');
        
        let startChar = charOffset;
        let startLine = lineIndex;

        // Scan upwards to capture decorators (lines starting with @)
        while (startLine > 0 && lines[startLine - 1].trim().startsWith('@')) {
          startLine--;
          startChar -= (lines[startLine].length + 1);
        }
        
        let endLine = lineIndex;

        // Scan forward to find end of block based on indentation
        for (let j = lineIndex + 1; j < lines.length; j++) {
          if (isLineEmptyOrComment(lines[j])) {
            endLine = j;
            continue;
          }
          const indent = getIndentationLevel(lines[j]);
          if (indent <= baseIndent) {
            break; // Block ended
          }
          endLine = j;
        }

        // Extract code
        const blockLines = lines.slice(startLine, endLine + 1);
        const symbolCode = blockLines.join('\n');
        const endChar = startChar + symbolCode.length;

        const dependencies = getIdentifiers(symbolCode).filter(id => id !== name && id !== 'def' && id !== 'class' && id !== 'self');

        // Extract declaration code
        let declCode = symbolCode;
        if (type === 'function') {
          const relativeDefIndex = lineIndex - startLine;
          let sigEndLine = relativeDefIndex;
          let parenCount = 0;
          for (let k = relativeDefIndex; k < blockLines.length; k++) {
            const bl = blockLines[k];
            for (const char of bl) {
              if (char === '(') parenCount++;
              else if (char === ')') parenCount--;
            }
            if (parenCount === 0 && bl.trim().endsWith(':')) {
              sigEndLine = k;
              break;
            }
          }
          const sigLines = blockLines.slice(0, sigEndLine + 1);
          const indentSpace = line.match(/^([ \t]*)/)?.[1] || '';
          declCode = sigLines.join('\n') + `\n${indentSpace}    pass`;
        } else if (type === 'class') {
          declCode = symbolCode
            .split('\n')
            .map(l => {
              if (l.trim().startsWith('def ')) {
                const indent = l.match(/^([ \t]*)/)?.[1] || '';
                return `${l.split(':')[0]}:\n${indent}    pass`;
              }
              return l;
            })
            .join('\n');
        }

        symbols.push({
          name,
          type,
          start: startChar,
          end: endChar,
          code: symbolCode,
          declCode,
          dependencies
        });
        
        lineIndex = endLine;
        charOffset = startChar + symbolCode.length + 1;
        continue;
      }
    }

    charOffset += lineLength;
  }

  return {
    filePath: absolutePath,
    imports,
    symbols
  };
}
